src/Campaigns/Blocks/CampaignDonorsBlock/render.php

```diff
@@ -0,0 +1,268 @@
+        }
+
+        /**
+         * @unreleased
+         */
+        public function render(): void
+        {
+            if ( ! $this->campaign) {
+                return;
+            }
+
+            $query = $this->buildDonorsQuery();
+            $donors = $this->formatDonorsData($query->getAll());
+            $this->renderBlockHtml($donors);
+        }
+
+        /**
+         * @unreleased
+         */
+        private function buildDonorsQuery(): CampaignDonationQuery
+        {
+            $sortBy = $this->attributes['sortBy'] ?? 'top-donors';
+            $query = new CampaignDonationQuery($this->campaign);
+
+            if ($sortBy === 'top-donors') {
+                $query->select(
+                    'donorIdMeta.meta_value as id',
+                    'SUM(amountMeta.meta_value) as amount',
+                    'donors.name as name'
+                )
+                    ->joinDonationMeta(DonationMetaKeys::DONOR_ID, 'donorIdMeta')
+                    ->joinDonationMeta(DonationMetaKeys::AMOUNT, 'amountMeta')
+                    ->leftJoin('give_donors', 'donorIdMeta.meta_value', 'donors.id', 'donors')
+                    ->groupBy('donorIdMeta.meta_value')
+                    ->orderBy('amount', 'DESC');
+            } else {
+                $query->select(
+                    'donation.ID as donationID',
+                    'donorIdMeta.meta_value as id',
+                    'companyMeta.meta_value as company',
+                    'donation.post_date as date',
+                    'amountMeta.meta_value as amount',
+                    'donors.name as name'
+                )
+                    ->joinDonationMeta(DonationMetaKeys::DONOR_ID, 'donorIdMeta')
+                    ->joinDonationMeta(DonationMetaKeys::COMPANY, 'companyMeta')
+                    ->joinDonationMeta(DonationMetaKeys::AMOUNT, 'amountMeta')
+                    ->leftJoin('give_donors', 'donorIdMeta.meta_value', 'donors.id', 'donors')
+                    ->orderBy('donation.ID', 'DESC');
+            }
+
+            $query->limit($this->attributes['donorsPerPage'] ?? 5);
+
+            if ( ! $this->attributes['showAnonymous']) {
+                $query->joinDonationMeta(DonationMetaKeys::ANONYMOUS, 'anonymousMeta')
+                    ->where('anonymousMeta.meta_value', '0');
+            }
+
+            return $query;
+        }
+
+        /**
+         * @unreleased
+         */
+        private function formatDonorsData(array $donors): array
+        {
+            return array_map(static function ($entry) {
+                if (isset($entry->date)) {
+                    $entry->date = Carbon::parse($entry->date)->diffForHumans();
+                }
+                $entry->amount = Money::fromDecimal($entry->amount, give_get_currency());
+
+                return $entry;
+            }, $donors);
+        }
+
+        /**
+         * @unreleased
+         */
+        private function renderBlockHtml($donors): void
```

Comment:
This is interesting that we are rendering the entire html in php, I was imagining the front-end to be a React instance.  I realize that's a pivotal shift at this point but something we should discuss for these blocks in the future.

Reply:
We are still beginning with the blocks. Maybe that discussion is better to happen sooner than later.